### PR TITLE
fix(client): Netlify using Node.js 20 should use rhel-openssl-3.0.x

### LINKED ISSUE #22244: Netlify deployment with Node 20 break Prisma (`Prisma Client could not locate the Query Engine for runtime "rhel-openssl-3.0.x"`)
As reported here: https://answers.netlify.com/t/new-builds-are-breaking-in-runtime-with-prisma-errors/107755/16

```
Prisma Client could not locate the Query Engine for runtime "rhel-openssl-3.0.x".
This happened because Prisma Client was generated for "debian-openssl-1.1.x", but the actual deployment required "rhel-openssl-3.0.x".
Add "rhel-openssl-3.0.x" to binaryTargets in the "schema.prisma" file and run prisma generate after saving it:
\n\ngenerator client {\n provider = "prisma-client-js"\n binaryTargets = ["native", "rhel-openssl-3.0.x"]\n}
```

> We tried adding more binary targets to the schema file binaryTargets = [“native”, “rhel-openssl-3.0.x”] but then the error changes to

```
Prisma Client could not locate the Query Engine for runtime "rhel-openssl-3.0.x".
\n\nThis is likely caused by tooling that has not copied "libquery_engine-rhel-openssl-3.0.x.so.node" to the deployment folder.
```
-----
When deploying on Netlify with Node 20 configured, the serverless function after deployment can not find the required query engine file. This is because the underlying image changed, and Prisma now defaults to OpenSSL3 instead of OpenSSL1.

### PR DESCRIPTION
**Pull request by Jolg42**
See [internal slack](https://prisma-company.slack.com/archives/C058VM009HT/p1701688209974089)

And testing https://prisma-company.slack.com/archives/CRGDU19K6/p1701770327881569

closes #22244

I tested the following with `5.7.0-integration-netlify-node20.4`

Via UI only
20 :white_check_mark: 
18 :white_check_mark: 
16 :white_check_mark: Note that that only changes the version for the build step, Node 18 is used for lambda.

Via `AWS_LAMBDA_JS_RUNTIME`
nodejs18.x and UI set to 18 :white_check_mark: 
nodejs18.x and UI set to 20 :white_check_mark: 
nodejs20.x and UI set to 18 :white_check_mark: 
nodejs20.x and UI set to 20 :white_check_mark:

### CODE REVIEW

File: packages/internals/src/get-generators/utils/getBinaryPathsByVersion.ts [hunk 1]
```diff
@@ -32,11 +32,24 @@ export async function getBinaryPathsByVersion({
...
-        value: 'rhel-openssl-1.0.x',
-      })
+    if (process.env.NETLIFY) {
+      const isRhelBinaryTarget1xInNeededVersions = neededVersion.binaryTargets.find(
+        (object) => object.value === 'rhel-openssl-1.0.x',
+      )
+      const isRhelBinaryTarget3xInNeededVersions = neededVersion.binaryTargets.find(
+        (object) => object.value === 'rhel-openssl-3.0.x',
+      )
+      if (parseInt(process.versions.node.split('.')[0]) >= 20 && !isRhelBinaryTarget3xInNeededVersions) {
```
[the remark this answers is not in the document]
Comment (Jolg42) [reply packages/internals/src/get-generators/utils/getBinaryPathsByVersion.ts:42]: Indeed https://docs.netlify.com/configure-builds/environment-variables/#netlify-configuration-variables

File: packages/client/src/generation/generateClient.ts [hunk 2]
```diff
@@ -300,10 +300,14 @@ export async function generateClient(options: GenerateClientOptions): Promise<vo
...
     for (const [binaryTarget, filePath] of Object.entries(enginePath)) {
       const fileName = path.basename(filePath)
-      const target =
-        process.env.NETLIFY && binaryTarget !== 'rhel-openssl-1.0.x'
-          ? path.join('/tmp/prisma-engines', fileName)
-          : path.join(finalOutputDir, fileName)
+      let target: string
+
+      if (process.env.NETLIFY && !['rhel-openssl-1.0.x', 'rhel-openssl-3.0.x'].includes(binaryTarget)) {
+        target = path.join('/tmp/prisma-engines', fileName)
```
[the remark this answers is not in the document]
Comment (Jolg42) [reply packages/client/src/generation/generateClient.ts:306]: I know because I went into git blame, but yes we need to clarify this with a comment

**Comment by ahmadardal** [comment]
Hello guys, we're suddenly having problems deploying our web app. This is the error

`"\nInvalid `prisma.offer.findUnique()` invocation:\n\n\nPrisma Client could not locate the Query Engine for runtime \"rhel-openssl-3.0.x\".\n\nThis happened because Prisma Client was generated for \"debian-openssl-1.1.x\", but the actual deployment required \"rhel-openssl-3.0.x\".\nAdd \"rhel-openssl-3.0.x\" to `binaryTargets` in the \"schema.prisma\" file and run `prisma generate` after saving it:\n\ngenerator client {\n  provider      = \"prisma-client-js\"\n  binaryTargets = [\"native\", \"rhel-openssl-3.0.x\"]\n}\n\nThe following locations have been searched:\n  /var/task/node_modules/.prisma/client\n  /var/task/node_modules/@prisma/client\n  /opt/build/repo/node_modules/@prisma/client\n  /tmp/prisma-engines",`

We're using prisma/client v5.3.1 with NodeJS 20. Is this related to your PR?

**Comment by janpio** [comment]
Yes @ahmadardal, see this issue: https://github.com/prisma/prisma/issues/22244
This PR is to fix the problem. Right now, stay on Node 18.

File: packages/client/src/generation/generateClient.ts [hunk 3]
```diff
@@ -300,10 +300,15 @@ export async function generateClient(options: GenerateClientOptions): Promise<vo
 
     for (const [binaryTarget, filePath] of Object.entries(enginePath)) {
       const fileName = path.basename(filePath)
-      const target =
-        process.env.NETLIFY && binaryTarget !== 'rhel-openssl-1.0.x'
-          ? path.join('/tmp/prisma-engines', fileName)
-          : path.join(finalOutputDir, fileName)
+      let target: string
+
+      // TODO comment
```
Comment (Jolg42) [inline packages/client/src/generation/generateClient.ts:305]:
```suggestion
      // Introduced in https://github.com/prisma/prisma/pull/6527
      // The engines that are not needed for the runtime deployment on AWS Lambda 
      // are moved to `/tmp/prisma-engines`
      // They will be ignored and not included in the final build, reducing its size
```

**Comment by Jolg42** [comment]
Note about what we discussed @millsp
The ecosystem-test is failing only because the expectation needs to be updated
```
-     "libquery_engine-rhel-openssl-1.0.x.so.node",
+     "libquery_engine-rhel-openssl-3.0.x.so.node",
```
See https://github.com/prisma/ecosystem-tests/actions/runs/7101395346/job/19329548093#step:8:886

File: packages/client/src/generation/utils/buildNFTAnnotations.ts [hunk 4]
```diff
@@ -30,8 +30,24 @@ export function buildNFTAnnotations(
...
+  // Add annotation for Netlify for a specific binaryTarget (depending on Node version and special env var)
   if (process.env.NETLIFY) {
-    platforms = ['rhel-openssl-1.0.x']
+    const isNodeMajor20OrUp = parseInt(process.versions.node.split('.')[0]) >= 20
+
+    // Netlify reads and changes the runtime version based on this env var
+    // https://docs.netlify.com/configure-builds/environment-variables/#netlify-configuration-variables
+    const awsRuntimeVersion = parseAWSNodejsRuntimeEnvVarVersion()
+    const isRuntimeEnvVar20OrUp = awsRuntimeVersion && awsRuntimeVersion >= 20
+    const isRuntimeEnvVar18OrDown = awsRuntimeVersion && awsRuntimeVersion <= 18
```
Comment (millsp) [inline packages/client/src/generation/utils/buildNFTAnnotations.ts:41]: Not critical: would be nice to have a shared function

**Review by millsp — APPROVED** [review]

File: packages/client/src/generation/utils/buildNFTAnnotations.ts (hunk 4 above)
Comment (Jolg42) [reply packages/client/src/generation/utils/buildNFTAnnotations.ts:41]: I'll merge this as is for now

**Pull request merged**